Implement MultiPayloadEnum support for projectEnumValue

tbkka opened the pull request:
This code rearchitects and simplifies the projectEnumValue support by introducing a new `TypeInfo` subclass for each kind of enum, including trivial, no-payload, single-payload, and multi-payload enums.  These classes are mostly private to the TypeLowering.cpp source file; the only change visible outside of that is a new base `EnumTypeInfo` that is a sibling of `RecordTypeInfo`.  Three of these new classes are used for MultiPayload enum support:

* "Unsupported" enums that we don't know how to project.  This TI returns "don't know" answers for any request to project an enum value or an XI but otherwise has correct layout and case data.

* "Simple" MP Enums that only use a separate tag value.  This includes dynamically-laid-out generic enums, as well as enums whose payloads have no spare bits.

* The general case of MP Enums that use spare bits, possibly in addition to a separate tag.  This logic can only be used, of course, if we can in fact compute a spare bit mask that agrees with the compiler.

The final challenge is to choose one of the above three handlings for every MPE.  In particular, there is still some work to robustly compute the spare bit mask for the third type above.

Resolves rdar://31666592

tbkka committed:
Implement MultiPayloadEnum support for projectEnumValue

This code rearchitects and simplifies the projectEnumValue support by
introducing a new `TypeInfo` subclass for each kind of enum, including trivial,
no-payload, single-payload, and three different classes for multi-payload enums:

* "UnsupportedEnum" that we don't understand.  This will allow us to
  return correct "don't know" answers in cases where the
  runtime lacks enough information to accurately handle
  a particular enum.

* MP Enums that only use a separate tag value.  This includes
  generic enums and other dynamic layouts, as well as
  enums whose payloads have no spare bits.

* MP Enums that use spare bits, possibly in addition to
  a separate tag.  This logic can only be used, of course,
  if we can in fact compute a spare bit mask that agrees
  with the compiler.

The final challenge is to choose one of the above three
handlings for every MPE.  The current code that computes
the spare bit mask for the third type above has some flaws
that need to be understood.

Next step: Figure out spare bit mask computation.  Any cases we can compute
correctly based on the information available at runtime can be handled by one of
the above three TI classes.  For the remaining cases, we'll probably have to
divert to "Unsupported" until we can arrange for the compiler to provide
us with augmented data.

tbkka commented:
This is "WIP" because there's still some work needed around the spare bit mask computation.  The basic implementation I have here is sufficient to test everything on x86_64 and 32-bit platforms.  In particular, there seems to be some question about whether visibility information that's only known to the compiler needs to be considered here.

There are a number of possible ways forward:
1. Query the memory reader to get target-specific data.  This would be a simple way to extend the current code to support all targets.  If that's sufficient, it would be an excellent answer.
2. Identify cases that cannot be handled correctly and divert them to `UnsupportedEnumTypeInfo` which fails all enum projection requests.  That will allow the cases that do work to become available to consumers.
3. Teach the compiler to add more metadata for use by the runtime so that the runtime can produce spare bit mask information for all cases.  Of course, I'd like to reduce the memory impact of this.

tbkka commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -286,132 +287,737 @@ bool RecordTypeInfo::readExtraInhabitantIndex(remote::MemoryReader &reader,
+
+  // Zero all bits except for the `n` most significant ones.
+  // XXX TODO: Big-endian support?
+  void keepOnlyMostSignificantBits(int n) {
```
Needed for computing the payload tag from the spare bits mask.  Note the comment about big-endian support.  I don't this is critical, since the corresponding compiler code lacks big-endian support as well.

tbkka commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -286,132 +287,737 @@ bool RecordTypeInfo::readExtraInhabitantIndex(remote::MemoryReader &reader,
+  template<typename IntegerType>
+  bool readMaskedInteger(remote::MemoryReader &reader,
+                         remote::RemoteAddress address,
+                         IntegerType *dest) const {
```
This uses a memory reader to gather bits from a payload into a single integer.

tbkka commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -286,132 +287,737 @@ bool RecordTypeInfo::readExtraInhabitantIndex(remote::MemoryReader &reader,
+
+// General multi-payload enum support for enums that do use spare
+// bits in the payload.
+class MultiPayloadEnumTypeInfo: public EnumTypeInfo {
```
MP enums that do not use spare bits can rely on the much simpler `SimpleMultiPayloadEnumTypeInfo` above.

tbkka commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -286,132 +287,737 @@ bool RecordTypeInfo::readExtraInhabitantIndex(remote::MemoryReader &reader,
+};
+
+// XXX FIXME XXX Make sure the following calculation returns `false` (failure)
+// when it is unable to accurately obtain the spare bit mask.
```
I'm not yet sure whether it's always possible for the runtime to do this calculation.  However, as long as this code is honest about it's capabilities and returns `false` when needed information is unavailable, everything else here should work.  The question is not really whether we can correctly support all MPEs, but what subset of MPEs we can correctly support.

tbkka replied:
Backed out for this PR.  A future PR will implement this by using additional data from the compiler.

tbkka commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -1295,12 +1896,39 @@ class EnumTypeInfoBuilder {
+        BitMask spareBitsMask(PayloadSize);
+        auto validSpareBitsMask = populateSpareBitsMask(Cases, spareBitsMask);
+
+        if (!validSpareBitsMask) {
```
Here's the critical logic that determines how to handle a particular MPE based on our current understanding.

tbkka commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -1522,21 +2144,19 @@ class LowerType
       return TC.getReferenceTypeInfo(Kind, ReferenceTI->getReferenceCounting());
 
+    if (auto *EnumTI = dyn_cast<EnumTypeInfo>(TI)) {
+      if (EnumTI->isOptional() && Kind == ReferenceKind::Weak) {
```
Note the previous version of this code just used `SubKind == SinglePayloadEnum` as a test for something being an Optional.  I've factored that out in order to make it easier to refine this decision.

tbkka reviewed (commented):
I've peppered some expository comments throughout the code to help orient reviewers.

jckarter commented:
Rather than have the runtime try to recompute spare bit information from the payload types, it might be simpler to encode the spare bits used by a particular enum in its metadata, since we never use spare bits for runtime type layout and always know them at compile time. Since this also means that the spare bit mask would never vary among different generic instantiations of a generic enum, we could add the information to the enum's type context descriptor, which is shared among all generic instantiations and lives in `__TEXT,__const`, which should reduce the runtime memory impact of the added metadata.

tbkka commented:
@jckarter  I understood that generic MPEs never used the spare bit strategy, so that part already doesn't vary with different instantiations.  @slavapestov also suggested adding spare bit info to the type metadata for fixed-layout MPEs that use it.  I'm not sure if that would be simpler or not; my current code to walk the type tree and accumulate spare bit info is only a couple dozen lines of code, though I need to do some more tests to see if I missed any cases.  My only real concern revolves around whether the compiler chooses MPE strategies based on information that's currently completely missing from the runtime metadata (such as visibility information).

jckarter commented:
I'm pretty sure the layout in its full generality does rely on information that's not available to the runtime. Encoding the spare bit info in the enum metadata would also be more robust if we decide to rev the layout algorithm for newly-defined enums eventually too.

tbkka commented:
"information that's not available to the runtime" -- Can you be more specific?  I'd like to have a test case to verify that the RemoteMirror code fails on cases that it should not be able to handle.

tbkka commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -1295,12 +1896,39 @@ class EnumTypeInfoBuilder {
-        // Dynamic multi-payload enums always use an extra tag to differentiate
-        // between cases
+        // Dynamic multi-payload enums cannot use spare bits, so they
+        // always use a separate tag value:
```
@jckarter @slavapestov -- Is my logic here correct:  If there's not a FixedDescriptor (according to the logic above), then this MPE does not use spare bits, so the `SimpleMultipayloadEnumTypeInfo` can handle it correctly.

jckarter replied:
Yeah, that should be the case that if the enum ever needs dynamic type layout, we don't use spare bits.

tbkka replied:
Thanks for confirming that.  I've just added a commit that keeps this case but diverts all fixed-layout MPEs to the `Unsupported` bucket.

tbkka commented on stdlib/public/Reflection/TypeLowering.cpp:
```diff
@@ -1295,12 +1896,39 @@ class EnumTypeInfoBuilder {
+          // Simple case that does not use spare bits
+          // This is correct as long as our local spare bits calculation
+          // above only returns an empty mask when the mask is really empty,
+          return TC.makeTypeInfo<SimpleMultiPayloadEnumTypeInfo>(
```
If I were to change this (and/or the case immediately below) to use `UnsupportedEnumTypeInfo`, I believe that would suffice to make this PR mergeable.  Some MPEs would not be projectable then, but that's inevitable even if we did make compiler changes, as this library will sometimes get used to inspect old binaries.  The important thing is that anything this library does project gets projected correctly.

tbkka replied:
Fixed-layout MPEs are now `Unsupported` so they won't be projected by the current code.  After I can add some mechanism for the compiler to expose accurate spare bit information, this can be revisited.

jckarter commented:
> "information that's not available to the runtime" -- Can you be more specific? I'd like to have a test case to verify that the RemoteMirror code fails on cases that it should not be able to handle.

To my best recollection, we get spare bits from the high byte of 64-bit object or function pointers, from the unused bits in enums, from padding between struct/tuple elements, and from the unused high bits of `Builtin.IntNN` fields when `NN` is not a power of two. Of those, I guess only the `Builtin.IntNN` case is strictly not reconstructable from runtime information. However, it seems like we would have to reimplement a lot of IRGen in order to be able to reconstruct the spare bit info for enum/struct padding.

tbkka committed:
Back out the experimental spare bits calculation and everything that depended on it

The experimental spare bits calculation helped to validate
the structure of this code and the handling of various
multi-payload enum details, but it cannot be done here
in a way that will always agree with the compiler.

So I've backed that out:  The result can still handle
certain multi-payload enums (those with dynamic layouts that
can never utilize spare bits) but the remainder will require
some additional work for the compiler to expose the real
spare bit data for use in the runtime.

tbkka commented:
@swift-ci Please test

tbkka commented:
I believe I've resolved the spare bits question sufficient to make this mergeable in its current state.  As such, I've removed the "WIP" from the PR title and plan to merge it as soon as it can pass tests.

tbkka commented:
@swift-ci Please test

tbkka committed:
Merge branch 'master' into tbkka-remoteMirror-projectEnum

tbkka commented:
@swift-ci Please test Linux

tbkka committed:
Explicitly include MetadataValues.h

tbkka commented:
@swift-ci Please test Linux

tbkka committed:
Avoid std::min with static const arguments

tbkka commented:
@swift-ci Please test

tbkka commented:
@swift-ci Please smoke test Linux

tbkka commented:
@swift-ci Please test Linux

tbkka commented:
@swift-ci Please test Linux
